### ISSUE TITLE
Kubernetes Skips Init Containers Beyond First During Some Pod Restarts

### ISSUE THREAD

**Issue by mike-capyh** [kubernetes/kubernetes#124002]
### What happened?

Init containers beyond the first are ignored during certain pod restarts.  In particular, when the pod sandbox suddenly temporarily disappears (possibly due to a node restart) the automatic restart sequence of the pod seems to skip init containers after the first.

### What did you expect to happen?

From the [Kubernetes docs](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#detailed-behavior):

"If the Pod [restarts](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#pod-restart-reasons), or is restarted, all init containers must execute again."

### How can we reproduce it (as minimally and precisely as possible)?

Below `repro.yaml` is a trimmed-down deployment:
two init containers and a primary container, all running bash.
Primary just sleeps, init containers each echo and exit.

Apply the yaml, let it get running, and then, eg, forcibly kill its pod sandbox (variety of methods, depending on CRI, host environment, etc..):

A specific repro that assumes a linux host, containerd:

```bash
kubectl create namespace delme
kubectl apply -f repro.yaml
kubectl get pods -n delme  # confirm the deployment started normally

# simulate a node/sandbox-level disruption by killing pod's slice (probably a better way to do this..?)
# run next command on **node host**:
sudo systemctl stop --force \
 "$(ps -o cgroup --pid "$(pgrep --full 'pgrepid:PkIOeQU4J6zI')" |
 sed -ne '2s/.*\(kubepods-besteffort-pod.*\.slice\).*/\1/p')"

kubectl describe pods -n delme
```

The restart count of the second init container is 0, while main container and first init container are 1.  Can repeat the stop, and the restart count of the first init container will increment, staying in sync with the main container and the overall pod restart counter, but the second init container stays at 0.  Start/finish time stamps in describe, and results of `kubectl logs`, are also consistent with the second init container not being re-run.

And here's `repro.yaml`:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: multi-init-container-bug-repro
  name: multi-init-container-bug-repro
  namespace: delme
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  selector:
    matchLabels:
      app: multi-init-container-bug-repro
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: multi-init-container-bug-repro
    spec:
      automountServiceAccountToken: false
      containers:
      - command:
        - /bin/bash
        - -c
        - sleep infinity && echo 'pgrepid:PkIOeQU4J6zI'
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: main
      initContainers:
      - command:
        - /bin/bash
        - -c
        - "echo 'init-container: First!'"
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: init-first
      - command:
        - /bin/bash
        - -c
        - "echo 'init-container: Second!'"
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: init-second
      restartPolicy: Always
```

## Practical Impact:

Deployments that use multiple init containers (which includes some popular ones, eg gitlab via helm) can get stuck during their auto restart after a transient node outage (requiring manual intervention to get unstuck).  In production, this is hopefully a rare (but potentially unpleasant) event.  It is fairly common and annoying in certain kinds of development/test configurations if eg intentional host node restarts are common.

## Workarounds:

Deleting the pod through `kubectl delete pod`, or `kubectl rollout restart`ing its eg deployment, causes init containers to be re-run when the pod is restarted (or more precisely, when the fresh replacement pod is started).

## Additional Notes

- [This issue](https://github.com/kubernetes/kubernetes/pull/47599) might be related.
- In the repro below, if instead of killing the whole pod sandbox, just the main container is killed, then `kubectl get pods` again reports a restart, but `describe` now shows *both* init containers are skipped.  This might be intended behavior? It is less-obviously-bad than arbitrarily skipping just some init containers, as happens in the main repro path.. but if the restart count in the output of `kubectl get pods` is to be interpreted as the "number of pod restarts", it seems like this path also contradicts the cited docs.

### Kubernetes version

Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.1

### Cloud provider

n/a (repros on bare-metal `kubeadm`)

### OS version

NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

### Install tools

kubeadm

### Container runtime (CRI) and version (if applicable)

containerd github.com/containerd/containerd v1.7.5 fe457eb99ac0e27b3ce638175ef8e68a7d2bc373

**Comment by rockswe** [comment]
Hi Mike.

I modified the init container commands to use the workaround script and added a shared emptyDir volume for the completion markers. This configuration adds an emptyDir volume named init-completion-markers, which is mounted into both init containers and the main container at /tmp/init-markers. Each init container now checks for the presence of its specific marker file before executing its main command. If the marker is not present, it executes the command and creates the marker. This makes sure that on pod restart, if the init container was previously completed successfully, it will not re-execute its payload, effectively working around the issue.

Try this revised version of "repro.yaml":

`apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: multi-init-container-bug-repro
  name: multi-init-container-bug-repro
  namespace: delme
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  selector:
    matchLabels:
      app: multi-init-container-bug-repro
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: multi-init-container-bug-repro
    spec:
      automountServiceAccountToken: false
      containers:
      - command:
        - /bin/bash
        - -c
        - sleep infinity && echo 'pgrepid:PkIOeQU4J6zI'
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: main
        volumeMounts:
        - name: init-completion-markers
          mountPath: /tmp/init-markers
      initContainers:
      - command:
        - /bin/bash
        - -c
        - >
          if [ ! -f /tmp/init-markers/init-first ]; then
            echo 'init-container: First!';
            touch /tmp/init-markers/init-first;
          fi
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: init-first
        volumeMounts:
        - name: init-completion-markers
          mountPath: /tmp/init-markers
      - command:
        - /bin/bash
        - -c
        - >
          if [ ! -f /tmp/init-markers/init-second ]; then
            echo 'init-container: Second!';
            touch /tmp/init-markers/init-second;
          fi
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: init-second
        volumeMounts:
        - name: init-completion-markers
          mountPath: /tmp/init-markers
      restartPolicy: Always
      volumes:
      - name: init-completion-markers
        emptyDir: {}
`

**Comment by mike-capyh** [comment]
Hi rockswe,

Thank you for taking the time to look into this, it is much appreciated!

I ran the repro procedure using the modified `repro.yaml`. Unfortunately, the second init container still gets skipped (restart count 0).

It appears that the proposed workaround is intended to prevent ill effects of init containers *running inconveniently often* (eg if have non-idempotent init).

However, the problem I ran into is that the init containers *don't run often enough* (K8s docs imply that all init containers *must* be re-run on Pod restart).

**Comment by chengjoey** [comment]
> ```shell
> # simulate a node/sandbox-level disruption by killing pod's slice (probably a better way to do this..?)
> # run next command on **node host**:
> sudo systemctl stop --force \
>  "$(ps -o cgroup --pid "$(pgrep --full 'pgrepid:PkIOeQU4J6zI')" |
>  sed -ne '2s/.*\(kubepods-besteffort-pod.*\.slice\).*/\1/p')"
> ```

`crictl pods` to get pod id and then `crictl stopp ${pod id}`

hi @mike-capyh , I would like to say that this is just a restart count error in the status, but in fact the second init container also restarts. You can increase the date command to display the time to verify.

```
- command:
        - /bin/bash
        - -c
        - "echo 'init-container: Second!' && date"
        image: debian:stable
        imagePullPolicy: IfNotPresent
        name: init-second
```

**Comment by mike-capyh** [comment]
Hi @chengjoey ,  thanks for looking into this, and for the `crictl` hint!

I get different behavior:  'unable to retrieve container logs for' (the second container, after the disruption).  This is with both the original method and the crictl method of causing the disruption.  Are you checking the `date` output via `kubectl logs`, or via some other method?

Example run on my end looks like:
```bash
# setup step, using `&& date` version of `repro.yaml`
pod=multi-init-container-bug-repro-6854f55d87-vrxrd

$ kubectl logs "$pod" -c init-first -n delme
init-container: First!
Tue Apr 30 17:49:48 UTC 2024

$ kubectl logs "$pod" -c init-second -n delme
init-container: Second!
Tue Apr 30 17:49:50 UTC 2024

# perform sandbox disruption step (via original or crictl method)

$ kubectl logs "$pod" -c init-first -n delme
init-container: First!
Tue Apr 30 17:51:32 UTC 2024

$ kubectl logs "$pod" -c init-second -n delme
unable to retrieve container logs for containerd://b7a9f8c62accb5092f78077e35d18a890e03d60d24b786bbeb2fb4b3e4bd9fca
```

**Comment by chengjoey** [comment]
> ```shell
> $ kubectl logs "$pod" -c init-second -n delme
> unable to retrieve container logs for containerd://b7a9f8c62accb5092f78077e35d18a890e03d60d24b786bbeb2fb4b3e4bd9fca
> ```

in my test, there is no such problem. If you can't see the log, you could try to view the container log directly `crictl ps` and `circtl logs ${container}`, you should be able to see that the second container is re-executed.

**Comment by mike-capyh** [comment]
Hi @chengjoey

As best I can tell, in my environment, `init-second` does not exist after the disruption.   I don't see it in `crictl ps`.  (details of my test run included below).  I'm wondering if there's a difference in our environments relevant to reproducing this issue.   What k8s / CRI versions are you using for your test runs?

```
$ crictl ps --all --pod "$pod"
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
185204e15f714       66d00b6e32f64       3 minutes ago       Running             main                0                   c17687b3bcfe1       multi-init-container-bug-repro-6854f55d87-82pk9
ba91ec2ecb82d       66d00b6e32f64       3 minutes ago       Exited              init-second         0                   c17687b3bcfe1       multi-init-container-bug-repro-6854f55d87-82pk9
68e8494ec2aea       66d00b6e32f64       3 minutes ago       Exited              init-first          0                   c17687b3bcfe1       multi-init-container-bug-repro-6854f55d87-82pk9

$ crictl logs 68e8494ec2aea
init-container: First!
Thu May  2 20:20:15 UTC 2024

$ crictl logs ba91ec2ecb82d
init-container: Second!
Thu May  2 20:20:15 UTC 2024

$ crictl stopp "$pod"
Stopped sandbox c17687b3bcfe1

pod='...' # grab new pod id from `crictl pods --namespace delme`

$ sudo crictl ps --all --pod "$pod"
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID              POD
6855a415096eb       66d00b6e32f64       4 minutes ago       Running             main                2                   e8132196dbed5       multi-init-container-bug-repro-6854f55d87-82pk9
2d6dab31efb4a       66d00b6e32f64       5 minutes ago       Exited              init-first          1                   e8132196dbed5       multi-init-container-bug-repro-6854f55d87-82pk9

# init-second appears to be missing

$ crictl logs 2d6dab31efb4a
init-container: First!
Thu May  2 20:26:43 UTC 2024

$ crictl ps --all --name 'init-second'  # no containers in output
$ crictl ps --all --name 'init-first'  # one container: 2d6dab31efb4a
```

**Comment by chengjoey** [comment]
@mike-capyh , I tested this demo in 1.29 version, and use minikube 

after `crictl stop ${podid}`:
```
CONTAINER           IMAGE                                                                             CREATED             STATE               NAME                      ATTEMPT             POD ID              POD
06f5b2f02aee1       busybox@sha256:6776a33c72b3af7582a5b301e3a08186f2c21a3409f0d2b52dfddbdbe24a5b04   9 minutes ago       Running             main                      1                   7e438a5884051       multi-init-container-bug-repro-654d8f4f9-7x9r7
5fa848b7acf62       busybox@sha256:6776a33c72b3af7582a5b301e3a08186f2c21a3409f0d2b52dfddbdbe24a5b04   9 minutes ago       Exited              init-second               0                   7e438a5884051       multi-init-container-bug-repro-654d8f4f9-7x9r7
79913e10fe6bc       busybox@sha256:6776a33c72b3af7582a5b301e3a08186f2c21a3409f0d2b52dfddbdbe24a5b04   9 minutes ago       Exited              init-first                1                   7e438a5884051       multi-init-container-bug-repro-654d8f4f9-7x9r7
```

the difference between `init-first` and `init-second` is the `attempt`

and I added some logs in kubelet, check the kubelet log:
```
May 06 02:47:47 minikube kubelet[1939964]: I0506 02:47:47.651497 1939964 pod_container_deletor.go:80] "Container not found in pod's containers" containerID="9107402c1f916453b0ddf8fc7a7ba4d054e14afad4d4811430a9a4bd772eeaa1"
May 06 02:48:17 minikube kubelet[1939964]: I0506 02:48:17.727624 1939964 scope.go:117] "RemoveContainer" containerID="0ad991f181f361c4fa3b4d9d754a61f61f98d26d4dfb17b1d35fe7f28a017d87"
May 06 02:48:17 minikube kubelet[1939964]: I0506 02:48:17.735222 1939964 scope.go:117] "RemoveContainer" containerID="e895782b40c5318528b4241f74cca1ec5ef2140b54a1469998fb82acf81001e7"
May 06 02:48:20 minikube kubelet[1939964]: I0506 02:48:20.228695 1939964 kuberuntime_container.go:209] Container status for init-first: &{{docker 0ad991f181f361c4fa3b4d9d754a61f61f98d26d4dfb17b1d35fe7f28a017d87} init-first exited 2024-05-06 02:44:49.828993715 +0000 UTC 2024-05-06 02:44:49.892845898 +0000 UTC 2024-05-06 02:44:49.893372815 +0000 UTC 0 busybox:latest docker-pullable://busybox@sha256:6776a33c72b3af7582a5b301e3a08186f2c21a3409f0d2b52dfddbdbe24a5b04  3460290366 0 0 Completed  <nil>}
May 06 02:48:20 minikube kubelet[1939964]: I0506 02:48:20.228762 1939964 kuberuntime_container.go:238] contianer: init-first, restartCount: 1
May 06 02:48:24 minikube kubelet[1939964]: I0506 02:48:24.426541 1939964 kuberuntime_container.go:209] Container status for init-second: <nil>
May 06 02:48:24 minikube kubelet[1939964]: I0506 02:48:24.426622 1939964 kuberuntime_container.go:238] contianer: init-second, restartCount: 0
May 06 02:48:25 minikube kubelet[1939964]: I0506 02:48:25.064745 1939964 scope.go:117] "RemoveContainer" containerID="c6036417154bb3408d6a36cf8d62530eb9378a2996ad7455366944a203539b48"
May 06 02:48:27 minikube kubelet[1939964]: I0506 02:48:27.101878 1939964 kuberuntime_container.go:209] Container status for main: &{{docker c6036417154bb3408d6a36cf8d62530eb9378a2996ad7455366944a203539b48} main exited 2024-05-06 02:44:55.499278448 +0000 UTC 2024-05-06 02:44:55.546857127 +0000 UTC 2024-05-06 02:48:17.684051033 +0000 UTC 137 busybox:latest docker-pullable://busybox@sha256:6776a33c72b3af7582a5b301e3a08186f2c21a3409f0d2b52dfddbdbe24a5b04  3514713412 0 0 Error  <nil>}
May 06 02:48:27 minikube kubelet[1939964]: I0506 02:48:27.101935 1939964 kuberuntime_container.go:238] contianer: main, restartCount: 1
```

see the comment,
https://github.com/kubernetes/kubernetes/blob/d83cd48e5ebbb1b073164a574ef9aa5a68569d9c/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1402-L1417

**Comment by mike-capyh** [comment]
It appears likely that changes in k8s `1.28.2` fixed the init container skip.

-with `minikube start`:   no issue
- with `minikube start --kubernetes-version 'v1.28.1'  issue repros!
- version bisection to find the last k8s version where the repro works:  turns out it is actually `1.28.1`.

Can you confirm that that the issue does occur with `--kubernetes-version 'v1.28.1'`?

If so, I'd be okay closing this as already fixed.

To answer "under what circumstances is this operation performed?":  that command line is just simplest repro I could find in my test environment, when I first ran into this issue.   Our actual use case was hardware restart of the host node (would restart the node, and most pods would recover normally, but multi-init-container pods tended to get stuck due to this issue).  Tearing out the sandbox directly seemed a faster/more-convenient way to reproduce the issue.

**Comment by chengjoey** [comment]
> Can you confirm that that the issue does occur with `--kubernetes-version 'v1.28.1'`?

Yes, the PR that fixes this problem has been merged in 1.28.2, the commit is[Feature-gate SidecarContainers code in pkg/kubelet/kuberuntime](https://github.com/kubernetes/kubernetes/commit/696f84aeb09740676031a1858711a9ac5700f8df)

/cc @gjkim42

**Comment by gjkim42** [comment]
Yes, this looks like the duplicate of https://github.com/kubernetes/kubernetes/issues/120247.

> This will address the #120247 with feature gate disabled.
> The case with feature gate enabled will be addressed by #120269 after this PR.

The bug has been fixed.

You can update the kubernetes and let me know if this happens again.

/close

**Comment by mike-capyh** [comment]
@chengjoey , @gjkim42 Ah, that makes sense.  I'm glad a fix is in place!  Thank you both for your time and help, it is much appreciated!
